src/main/java/dev/emortal/velocity/privatemessages/PrivateMessageListener.java

```diff
@@ -1,34 +1,41 @@
 package dev.emortal.velocity.privatemessages;
 
+import com.velocitypowered.api.proxy.Player;
 import com.velocitypowered.api.proxy.ProxyServer;
 import dev.emortal.api.message.messagehandler.PrivateMessageCreatedMessage;
 import dev.emortal.api.model.messagehandler.PrivateMessage;
-import dev.emortal.velocity.messaging.MessagingCore;
+import dev.emortal.velocity.messaging.MessagingModule;
 import net.kyori.adventure.text.minimessage.MiniMessage;
 import net.kyori.adventure.text.minimessage.tag.resolver.Placeholder;
 import org.jetbrains.annotations.NotNull;
 
 import java.util.UUID;
 
-public class PrivateMessageListener {
+public final class PrivateMessageListener {
     private static final MiniMessage MINI_MESSAGE = MiniMessage.miniMessage();
 
     private static final String PRIVATE_MESSAGE_RECEIVED_FORMAT = "<dark_purple>(<light_purple><username> -> You<dark_purple>) <light_purple><message>";
 
 
-    public PrivateMessageListener(@NotNull ProxyServer proxy, @NotNull MessagingCore messaging, LastMessageCache lastMessageCache) {
-        messaging.addListener(PrivateMessageCreatedMessage.class, message -> {
-            PrivateMessage privateMessage = message.getPrivateMessage();
+    private final ProxyServer proxy;
+    private final LastMessageCache lastMessageCache;
 
-            proxy.getPlayer(UUID.fromString(privateMessage.getRecipientId())).ifPresent(player -> {
-                player.sendMessage(MINI_MESSAGE.deserialize(PRIVATE_MESSAGE_RECEIVED_FORMAT,
-                        Placeholder.parsed("username", privateMessage.getSenderUsername()),
-                        Placeholder.unparsed("message", privateMessage.getMessage())
-                ));
+    public PrivateMessageListener(@NotNull ProxyServer proxy, @NotNull MessagingModule messaging, LastMessageCache lastMessageCache) {
+        this.proxy = proxy;
+        this.lastMessageCache = lastMessageCache;
 
-                lastMessageCache.setLastMessage(player.getUniqueId(), privateMessage.getSenderUsername());
-                lastMessageCache.setLastMessage(UUID.fromString(privateMessage.getSenderId()), player.getUsername());
-            });
-        });
+        messaging.addListener(PrivateMessageCreatedMessage.class, message -> this.onPrivateMessageCreated(message.getPrivateMessage()));
+    }
+
+    private void onPrivateMessageCreated(@NotNull PrivateMessage message) {
+        Player player = this.proxy.getPlayer(UUID.fromString(message.getRecipientId())).orElse(null);
+        if (player == null) return;
+
+        player.sendMessage(MINI_MESSAGE.deserialize(PRIVATE_MESSAGE_RECEIVED_FORMAT,
+                Placeholder.parsed("username", message.getSenderUsername()),
+                Placeholder.unparsed("message", message.getMessage())));
+
+        this.lastMessageCache.setLastMessage(player.getUniqueId(), message.getSenderUsername());
+        this.lastMessageCache.setLastMessage(UUID.fromString(message.getSenderId()), player.getUsername());
     }
```

Comment:
I imagine this was an issue before but:

We still need to update the last message cache for one player even if the other is missing (i.e they are on different proxies). So wecant have the fail fast of player == null then return as we still have to update last message if senderId is on this proxy.

Reply:
As the recipient username is only gotten from the proxy we will need to either:

A) message handler resolves the username of the recipient. Technically the sending server can pass this to it as the sending server resolves this to send the message.

B) have the proxy resolve this.